Source/WebCore/Modules/screen-wake-lock/WakeLockManager.cpp

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (C) 2022 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "WakeLockManager.h"
+
+#include "Document.h"
+#include "SleepDisabler.h"
+#include "WakeLockSentinel.h"
+
+namespace WebCore {
+
+WakeLockManager::WakeLockManager(Document& document)
+    : m_document(document)
+{
+    m_document.registerForVisibilityStateChangedCallbacks(*this);
+}
+
+WakeLockManager::~WakeLockManager()
+{
+    m_document.unregisterForVisibilityStateChangedCallbacks(*this);
+}
+
+void WakeLockManager::addWakeLock(Ref<WakeLockSentinel>&& lock)
+{
+    auto type = lock->type();
+    auto& locks = m_wakeLocks.ensure(type, [] { return Vector<RefPtr<WakeLockSentinel>>(); }).iterator->value;
+    ASSERT(!locks.contains(lock.ptr()));
+    locks.append(WTFMove(lock));
+
+    if (locks.size() != 1)
+        return;
+
+    switch (type) {
+    case WakeLockType::Screen:
```

Comment:
It seems silly and inefficient to have a WakeLockType enum with a single value. I realize that this is how the specification is defined, but it is strange! Why pay the cost of a switch/case/branch here when it can never be anything else?

Reply:
I wrote it this way to match the spec and for future extensibility as different types of lock can be supported (I think Blink has a system lock for e.g.). Hopefully the compiler is smart enough to optimize this out but if it isn't, this is not hot code at all.